### PR TITLE
Fix multi-platform soci standalone convert

### PR DESCRIPTION
**Pull request by prafgup**
**Issue #, if available:**

Fixes - https://github.com/awslabs/soci-snapshotter/issues/1948

**Description of changes:**

The standalone layout loader parsed only index.json.manifests[0], so any input whose top-level index listed multiple per-platform manifests directly (e.g. from go-containerregistry's layout.Write) lost every platform after the first silently when using --all-platforms, or with a misleading "manifest not found" when using --platform.

Replace the single-element parse with resolveLayoutRoot, which handles all three shapes index.json (single image, nested manifest list, flat per-platform list), filters out children whose blobs are missing, and promotes the top-level index to a content-addressable blob when it is itself the manifest list.

Running the reproduction steps here now gives correct results - https://github.com/awslabs/soci-snapshotter/issues/1948

```
SOCI_BIN=/Users/prafulg/projects/prafulg/soci-snapshotter/out/soci ./reproduce.sh


=== INPUT platforms ===
{
  "platform": {
    "architecture": "arm64",
    "os": "linux"
  },
  "digest": "sha256:7094b474ca994e198cf92414e0c552a315eb48237053292756425c1f24e819ce"
}
{
  "platform": {
    "architecture": "amd64",
    "os": "linux"
  },
  "digest": "sha256:965a4be6edf9b4658f854d042d21f92fc4158871547fc769be7a1150cc677ad8"
}
{
  "platform": {
    "architecture": "arm",
    "os": "linux",
    "variant": "v7"
  },
  "digest": "sha256:2381361da9d2c687488804c23e41b3aaf99101609d6d985b9c4b8f26eb5517f6"
}

=== convert --all-platforms ===
layer sha256:fee30871cd8cc89f6c3e6ed53bede46c1f08bb1d8051cd6384517d756099d454 -> ztoc sha256:34d0ce07a2d20dc49ac1459cf1448d2a2dafc1f9660f4d149ac44b6475333cdd
layer sha256:552ed30d0b0dfa9ae77aab6a08c2e1dbc8b3cef6243e9c9fc9715f566b08f202 -> ztoc sha256:4396be51bf34bb3d58192c7a4adb355585e4fbe770c3248949d4b235a9b76122
layer sha256:9fa8a8506df39d5a95b9b2e12bd982f79f31a71e8b238b757fc04e3b53e06673 -> ztoc sha256:8c5a21917a338c59324b0c66d1f5bdadf3342a5a6a76c7309da9cbe4ed9bb869
(exit 0)
{
  "platform": {
    "architecture": "arm64",
    "os": "linux"
  },
  "artifactType": null
}
{
  "platform": {
    "architecture": "amd64",
    "os": "linux"
  },
  "artifactType": null
}
{
  "platform": {
    "architecture": "arm",
    "os": "linux",
    "variant": "v7"
  },
  "artifactType": null
}
{
  "platform": {
    "architecture": "arm64",
    "os": "linux"
  },
  "artifactType": "application/vnd.amazon.soci.index.v2+json"
}
{
  "platform": {
    "architecture": "amd64",
    "os": "linux"
  },
  "artifactType": "application/vnd.amazon.soci.index.v2+json"
}
{
  "platform": {
    "architecture": "arm",
    "os": "linux",
    "variant": "v7"
  },
  "artifactType": "application/vnd.amazon.soci.index.v2+json"
}

=== convert --platform linux/amd64 (not first) ===
layer sha256:552ed30d0b0dfa9ae77aab6a08c2e1dbc8b3cef6243e9c9fc9715f566b08f202 -> ztoc sha256:4396be51bf34bb3d58192c7a4adb355585e4fbe770c3248949d4b235a9b76122
(exit 0)
{
  "platform": {
    "architecture": "arm64",
    "os": "linux"
  },
  "artifactType": null
}
{
  "platform": {
    "architecture": "amd64",
    "os": "linux"
  },
  "artifactType": null
}
{
  "platform": {
    "architecture": "arm",
    "os": "linux",
    "variant": "v7"
  },
  "artifactType": null
}
{
  "platform": {
    "architecture": "amd64",
    "os": "linux"
  },
  "artifactType": "application/vnd.amazon.soci.index.v2+json"
}
```

**Testing performed:**

Added Integration test for multiple platform image.

```
lima sudo GO_TEST_FLAGS="-run TestStandalone -count=1" make integration 2>&1 | tail -25)
  ⎿  === RUN   TestStandaloneConvertBasic/tar-to-tar
     --- PASS: TestStandaloneConvertBasic/tar-to-tar (0.22s)
     === RUN   TestStandaloneConvertBasic/dir-to-dir
     --- PASS: TestStandaloneConvertBasic/dir-to-dir (0.28s)
     === RUN   TestStandaloneConvertBasic/tar-to-dir
     --- PASS: TestStandaloneConvertBasic/tar-to-dir (0.27s)
     --- PASS: TestStandaloneConvertBasic (24.86s)
     === RUN   TestStandaloneConvertSpecificPlatform
     --- PASS: TestStandaloneConvertSpecificPlatform (21.11s)
     === RUN   TestStandaloneInvalidConversion
     === RUN   TestStandaloneInvalidConversion/nonexistent_input
     --- PASS: TestStandaloneInvalidConversion/nonexistent_input (0.03s)
     === RUN   TestStandaloneInvalidConversion/invalid_format
     --- PASS: TestStandaloneInvalidConversion/invalid_format (0.03s)
     === RUN   TestStandaloneInvalidConversion/missing_destination
     --- PASS: TestStandaloneInvalidConversion/missing_destination (0.03s)
     --- PASS: TestStandaloneInvalidConversion (1.58s)
     === RUN   TestStandaloneConvertIdempotent
     --- PASS: TestStandaloneConvertIdempotent (20.65s)
     === RUN   TestStandaloneConvertAllPlatforms
     --- PASS: TestStandaloneConvertAllPlatforms (38.93s)
     PASS
     ok         github.com/awslabs/soci-snapshotter/integration 200.444s
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by prafgup** [comment]
Hey @sondavidb @Shubhranshu153 @coderbirju can you take a look at this fix 😄 .
(Also would appreciate a minor release if possible so we can consume it directly)

File: cmd/soci/commands/internal/standalone.go [hunk 1]
```diff
@@ -70,22 +70,11 @@ func LoadImage(ctx context.Context, inputPath string, tmpDir string) (*Standalon
 	if err != nil {
 		return nil, fmt.Errorf("failed to read index.json from %s: %w", inputPath, err)
 	}
-	rootDesc, err := parseRootDescriptor(indexData)
+	rootDesc, err := resolveLayoutRoot(tmpDir, indexData)
```
Comment (KlwntSingh) [inline cmd/soci/commands/internal/standalone.go:73]: Can you please add sample image spec for both `nerdctl save` and `go-containerregistry` to compare the structure of image specs supported by standaLone?

File: cmd/soci/commands/internal/standalone.go [hunk 2]
```diff
@@ -139,73 +128,81 @@ func SaveImageToDir(srcDir string, desc ocispec.Descriptor, outputPath string) e
...
-		return ocispec.Descriptor{}, errors.New("manifest list contains no manifests with available blobs")
+	if mediaType == "" {
+		mediaType = ocispec.MediaTypeImageIndex
 	}
 
-	// If all platforms are available, keep the original manifest list
-	if len(available) == len(manifestList.Manifests) {
-		return listDesc, nil
+	var list ocispec.Index
+	if err := json.Unmarshal(listBytes, &list); err != nil {
```
Comment (KlwntSingh) [inline cmd/soci/commands/internal/standalone.go:171]: The assumption is that `listBytes` is of `index` media type but that is wrong. It can be index or list of manifests. The correct way is to iterate over the list to see if it is index or manifest and then do corresponding action.

Comment (KlwntSingh) [reply cmd/soci/commands/internal/standalone.go:171]: If it is a index, then the current implementation makes sense.
If it is a list of manifest, then return the list of descriptors representing each manifest.

Comment (prafgup) [reply cmd/soci/commands/internal/standalone.go:171]: - `application/vnd.oci.image.index.v1+json` and `application/vnd.docker.distribution.manifest.list.v2+json` have similar JSON `({schemaVersion, mediaType, manifests: [...]})`
- Both unmarshal cleanly into ocispec.Index :)

File: cmd/soci/commands/internal/standalone.go (hunk 1 above)
Comment (prafgup) [reply cmd/soci/commands/internal/standalone.go:73]: For nerdctl - 

```
sudo nerdctl save --all-platforms -o /tmp/n.tar docker.io/library/alpine:3.19
mkdir -p /tmp/nerdctl-layout && tar -xf /tmp/n.tar -C /tmp/nerdctl-layout
jq . /tmp/nerdctl-layout/index.json



{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.index.v1+json",
      "digest": "sha256:6baf43584bcb78f2e5847d1de515f23499913ac9f12bdf834811a3145eb11ca1",
      "size": 8077,
      "annotations": {
        "containerd.io/distribution.source.docker.io": "library/alpine",
        "io.containerd.image.name": "docker.io/library/alpine:3.19",
        "org.opencontainers.image.ref.name": "3.19"
      }
    }
  ]
}


DGST=$(jq -r '.manifests[0].digest' /tmp/nerdctl-layout/index.json | sed s/sha256://)
jq . /tmp/nerdctl-layout/blobs/sha256/$DGST


{
  "manifests": [
    {
      "annotations": {
        "com.docker.official-images.bashbrew.arch": "amd64",
        "org.opencontainers.image.base.name": "scratch",
        "org.opencontainers.image.created": "2025-10-08T21:19:15Z",
        "org.opencontainers.image.revision": "ee939d52345248420cf62d4606ccc7152bc5a107",
        "org.opencontainers.image.source": "https://github.com/alpinelinux/docker-alpine.git#ee939d52345248420cf62d4606ccc7152bc5a107:x86_64",
        "org.opencontainers.image.url": "https://hub.docker.com/_/alpine",
        "org.opencontainers.image.version": "3.19.9"
      },
      "digest": "sha256:b58899f069c47216f6002a6850143dc6fae0d35eb8b0df9300bbe6327b9c2171",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      },
      "size": 1022
    },
    {
      "annotations": {
        "com.docker.official-images.bashbrew.arch": "amd64",
        "vnd.docker.reference.digest": "sha256:b58899f069c47216f6002a6850143dc6fae0d35eb8b0df9300bbe6327b9c2171",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "digest": "sha256:ea72def7a47260a6b8cea592a2d0e07b36131d7e0694cfa3417615f87b3ae30d",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      },
      "size": 838
    },
    {
      "annotations": {
        "com.docker.official-images.bashbrew.arch": "arm32v6",
        "org.opencontainers.image.base.name": "scratch",
....
....
      "platform": {
        "architecture": "s390x",
        "os": "linux"
      },
      "size": 1021
    },
    {
      "annotations": {
        "com.docker.official-images.bashbrew.arch": "s390x",
        "vnd.docker.reference.digest": "sha256:5857a3de06e5ebf42101b7c3ea2c72acc06f4e95636191e23a5f4a328d0f9b2f",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "digest": "sha256:5d8c15d788da6df4c7664a81799d6ab333ca4b7da66448f687e07951dfc5e3f7",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      },
      "size": 838
    }
  ],
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "schemaVersion": 2
}

```

For go-containerregistry - 

```
sudo go run ./gcr-write.go docker.io/library/alpine:3.19 /tmp/gcr-layout
jq . /tmp/gcr-layout/index.json

{
  "manifests": [
    {
      "annotations": {
        "com.docker.official-images.bashbrew.arch": "amd64",
        "org.opencontainers.image.base.name": "scratch",
        "org.opencontainers.image.created": "2025-10-08T21:19:15Z",
        "org.opencontainers.image.revision": "ee939d52345248420cf62d4606ccc7152bc5a107",
        "org.opencontainers.image.source": "https://github.com/alpinelinux/docker-alpine.git#ee939d52345248420cf62d4606ccc7152bc5a107:x86_64",
        "org.opencontainers.image.url": "https://hub.docker.com/_/alpine",
        "org.opencontainers.image.version": "3.19.9"
      },
      "digest": "sha256:b58899f069c47216f6002a6850143dc6fae0d35eb8b0df9300bbe6327b9c2171",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      },
      "size": 1022
    },
    {
      "annotations": {
        "com.docker.official-images.bashbrew.arch": "amd64",
        "vnd.docker.reference.digest": "sha256:b58899f069c47216f6002a6850143dc6fae0d35eb8b0df9300bbe6327b9c2171",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "digest": "sha256:ea72def7a47260a6b8cea592a2d0e07b36131d7e0694cfa3417615f87b3ae30d",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {
....

      },
      "digest": "sha256:5d8c15d788da6df4c7664a81799d6ab333ca4b7da66448f687e07951dfc5e3f7",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      },
      "size": 838
    }
  ],
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "schemaVersion": 2

```

```

cat ./gcr-flat.go
---


package main

import (
	"fmt"
	"os"

	"github.com/google/go-containerregistry/pkg/name"
	v1remote "github.com/google/go-containerregistry/pkg/v1/remote"
	"github.com/google/go-containerregistry/pkg/v1/layout"
)

func main() {
	if len(os.Args) != 3 {
		fmt.Fprintln(os.Stderr, "usage: gcr-write <ref> <outdir>")
		os.Exit(1)
	}
	ref, err := name.ParseReference(os.Args[1])
	if err != nil { panic(err) }
	desc, err := v1remote.Get(ref)
	if err != nil { panic(err) }
	idx, err := desc.ImageIndex()
	if err != nil { panic(err) }
	if _, err := layout.Write(os.Args[2], idx); err != nil { panic(err) }
	fmt.Println("wrote", os.Args[2])
}
```

So basically now we support both index and manifest list based index :) 
(also added output of the previously failing minimal script from https://github.com/awslabs/soci-snapshotter/issues/1948 in the description that shows that now the image is correctly indexed + integration test verifies it as well)

**Review by prafgup — COMMENTED** [review]
@KlwntSingh replied to your comments 😄 
Also the the new integration test uses nerdctl to pull and verify all-platforms, previously this integration test would have failed, while other integration tests are untouched in the standalone path which I added previously as a part of https://github.com/awslabs/soci-snapshotter/pull/1881

**Comment by prafgup** [comment]
Pinging @sondavidb regrading - https://github.com/awslabs/soci-snapshotter/pull/1943#issuecomment-4359313268

File: cmd/soci/commands/internal/standalone.go [hunk 3]
```diff
@@ -139,73 +128,81 @@ func SaveImageToDir(srcDir string, desc ocispec.Descriptor, outputPath string) e
...
-	var index ocispec.Index
-	if err := json.Unmarshal(indexData, &index); err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("failed to unmarshal index.json: %w", err)
-	}
-	if len(index.Manifests) == 0 {
+// resolveLayoutRoot returns a descriptor for the OCI image layout's root manifest.
+// It accepts index.json shapes produced by common tools: a single image manifest,
+// a single descriptor pointing at a nested manifest list (e.g. nerdctl save), or
+// a flat list of per-platform manifests (e.g. go-containerregistry layout.Write).
+// Children whose blobs are missing are filtered out.
```
Comment (sondavidb) [inline cmd/soci/commands/internal/standalone.go]: Is this a common scenario? I think it's good practice regardless as a sanity check but wondering if you've seen this before.

File: cmd/soci/commands/internal/standalone.go [hunk 4]
```diff
@@ -139,73 +128,81 @@ func SaveImageToDir(srcDir string, desc ocispec.Descriptor, outputPath string) e
...
 }
 
-// parseRootDescriptor unmarshals OCI index JSON and returns the manifest descriptor.
-func parseRootDescriptor(indexData []byte) (ocispec.Descriptor, error) {
-	var index ocispec.Index
-	if err := json.Unmarshal(indexData, &index); err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("failed to unmarshal index.json: %w", err)
-	}
-	if len(index.Manifests) == 0 {
+// resolveLayoutRoot returns a descriptor for the OCI image layout's root manifest.
```
Comment (sondavidb) [inline cmd/soci/commands/internal/standalone.go]: To be clear, the idea of this function is to normalize the input to a manifest list in a descriptor?

File: cmd/soci/commands/internal/standalone.go [hunk 5]
```diff
@@ -139,73 +128,81 @@ func SaveImageToDir(srcDir string, desc ocispec.Descriptor, outputPath string) e
...
-	// Find which platform manifests have their blobs present
-	var available []ocispec.Descriptor
-	for _, desc := range manifestList.Manifests {
-		p := filepath.Join(layoutDir, "blobs", desc.Digest.Algorithm().String(), desc.Digest.Encoded())
-		if _, err := os.Stat(p); err == nil {
-			available = append(available, desc)
+	// Locate the manifest list to walk. Either index.json points at a nested list
+	// blob, or index.json is itself the list.
+	var (
+		listDesc  ocispec.Descriptor
```
Comment (sondavidb) [inline cmd/soci/commands/internal/standalone.go]: Looks like this is only ever populated when len(top.Manifests) == 1, and is never modified. Can we remove this variable and just check for len(top.Manifests) in 185? Feels like that would make more sense to me given that we store an entire descriptor.

**Comment by prafgup** [comment]
Seems like one of the checks failed 🤔 "[Pre-build / check (pull_request)](https://github.com/awslabs/soci-snapshotter/actions/runs/25026994856/job/73987126539?pr=1947)[Pre-build / check (pull_request)](https://github.com/awslabs/soci-snapshotter/actions/runs/25026994856/job/73987126539?pr=1947)" 


Might be due to me not having `Signed-off-by: Praful Gupta <prafulgupta6@gmail.com>` in the commit 🤔 can add that once the build check succeeds.

**Review by sondavidb — COMMENTED** [review]
Will look at the test a little later

File: cmd/soci/commands/internal/standalone.go (hunk 3 above)
Comment (prafgup) [reply cmd/soci/commands/internal/standalone.go]: Yes, in the initial testing using nerdctl save and crane pull, if I download multi-platform image with a given target platform the full index.json was written but only the blobs of the requested platforms were present. Thus walking into a child manifest whose blobs are missing from the dir/tar errored it out, so filtering this out.

File: cmd/soci/commands/internal/standalone.go (hunk 4 above)
Comment (prafgup) [reply cmd/soci/commands/internal/standalone.go]: Yes, it returns the canonical root descriptor for the layout. The output could be a single image manifest or a manifest list descriptor. Have made it more clear in the [recent push](https://github.com/awslabs/soci-snapshotter/compare/39a613bc97911e47fba2041a531793bb60a21250..fccc6c42c90864608c8161af031cad036dfee216) :)

File: cmd/soci/commands/internal/standalone.go (hunk 5 above)
Comment (prafgup) [reply cmd/soci/commands/internal/standalone.go]: Makes sense, changed it :)

**Review by prafgup — COMMENTED** [review]
Addressed comments :) and added the signed-off in the commit.

File: integration/convert_standalone_test.go
Comment (sondavidb) [inline integration/convert_standalone_test.go:1]: The test itself looks good. Wondering if we could add a test case for the missing blobs case? (Which it seems is what caused you to discover this bug to begin with?) If it's too complex I'm OK skipping it, it seems to be more of an edge case anyway, but it might be useful. (Maybe a unit test for this too.)

Otherwise, LGTM!

Comment (prafgup) [reply integration/convert_standalone_test.go:1]: (Replying from phone)
iirc the normal single platform convert with crane/nerdclt pull with multi-platform image in integration test should cover it already. I remember those test failing without this filtering logic.

I can confirm later and maybe add unit/integration test as follow-ups.

**Review by sondavidb — APPROVED** [review]
Thanks for the fix!

**Comment by prafgup** [comment]
Hey maintainers 🙂! Can someone else review/merge this (https://github.com/awslabs/soci-snapshotter/pull/1947) and https://github.com/awslabs/soci-snapshotter/pull/1943 (already approved by @\sondavidb) , and potentially release a minor version this with the critical vulnerability fix (already merged) - https://github.com/awslabs/soci-snapshotter/pull/1953 , would really help in consuming this package from releases on our end.

(cc.ing a couple devs to take a look as well) @KlwntSingh @Swapnanil-Gupta @coderbirju

**Review by Swapnanil-Gupta — APPROVED** [review]

**Comment by Swapnanil-Gupta** [comment]
Thanks.

**Comment by prafgup** [comment]
Resolved comments :) for merge compatibility ! 

Thanks for the review !
